"Freeze" a room when the last admin of that room leaves

anoadragon453 opened the pull request:
If the last admin of a room departs, and thus the room no longer has any admins within it, we "freeze" the room. Freezing a room means that the power level required to do anything in the room (sending messages, inviting others etc) will require power level 100.

At the moment, an admin can come back and unfreeze the room manually. The plan is to eventually make unfreezing of the room automatic on admin rejoin, though that will be in a separate PR.

This *could* work in mainline, however if the admin who leaves is on a homeserver without this functionality, then the room isn't frozen. I imagine this would probably be pretty confusing to people. Part of this feature was allowing Synapse modules to send events, which has been implemented in mainline at  https://github.com/matrix-org/synapse/pull/8479, and cherry-picked to the `dinsic` fork in 62c7b10. The actual freezing logic has been implemented here in the RoomAccessRules module.

Reviewable commit-by-commit, commit messages have the breakdown.

babolivier commented:
I know it makes things more complicated, but I think this PR should be split in two: one on mainline Synapse to add the capability to send an event to the third party event rules, and one here built on top if it to implement this specific behaviour. Otherwise we'll end up diverging from mainline's API or creating merge sadness.

anoadragon453 commented:
@babolivier When you say built on top, are you recommending a merge from mainline, or just cherry-picking the commit?

babolivier commented:
> @babolivier When you say built on top, are you recommending a merge from mainline, or just cherry-picking the commit?

Ideally merging from mainline. If that's not possible then maybe a merge from the feature branch, or cherry-picking.

anoadragon453 commented:
@babolivier OK, I'll break this up then.

anoadragon453 commented on synapse/events/third_party_rules.py:
```diff
@@ -12,9 +12,67 @@
+
+class EventCreator:
+    def __init__(self, hs):
+        self.event_creation_handler = hs.get_event_creation_handler()
```
This should be `self._event_creation_handler` to discourage third party modules from reaching in and using it.

anoadragon453 committed:
Add room freeze functionality when the last admin leaves

When a membership event is checked by the room access rules, we check if it's both a
leave membership, as well as whether this belongs to the last admin (PL>=100) user in
a room. If it does, we freeze the room.

When freezing the room, we take the existing power level state event of the room, modify
it so that everything requires PL100 (meaning the admin could come back and fix the room
if desired), sends the state event, and then actually leaves.

We are careful to only do this if the user leaving is a local user. Otherwise, we wouldn't
be able to puppet them and send a state event into the room. To enable this, we also now
pass server_name to room access rules.

anoadragon453 committed:
Fix new power level event being deduplicated, thus not persisted

Turns out we were overridding the content of the old power level event in the room, which
means that when it came time to persist the new power level event, the code compared the
content of the two and determined that the new state event was a duplicated, and thus it
wasn't persisted - so clients didn't see it down sync, and the new power levels were not
enforced after a server restart.

Copying the old power levels dict instead of using it directly solves the issue.

anoadragon453 committed:
Changelog

anoadragon453 committed:
Tests for room freezing

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -446,8 +449,130 @@ def _on_membership_or_invite(
+        admin_users = [
+            user
+            for user, power_level in power_level_content["users"].items()
+            if power_level >= 100
```
we're sure this is how we want to define "admin" ?

anoadragon453 replied:
I asked the same, but @giomfo mentioned in our internal room that yes, this is what we'd like to do.

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -446,8 +449,130 @@ def _on_membership_or_invite(
+        # Get every admin user defined in the room's state
+        admin_users = [
+            user
+            for user, power_level in power_level_content["users"].items()
```
it feels a bit asymmetric to check that `power_level_content` is a dict, but not `power_level_content["users"]`.

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -446,8 +449,130 @@ def _on_membership_or_invite(
+            return
+
+        # Get every admin user defined in the room's state
+        admin_users = [
```
it might be more efficient to make this a set:

```suggestion
        admin_users = {
```

(and the corresponding change below)

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -446,8 +449,130 @@ def _on_membership_or_invite(
+                power_level_content[key] = 100
+        power_level_content["events"] = {}
+
+        # Freeze the room by raising the required power level to send events to 100
+        logger.info("Freezing room '%s'", event.room_id)
```
these probably belong further up, at line 507ish?

anoadragon453 replied:
Fair yeah. I originally moved it down as it was firing even if the conditions were not met for "the last admin leaving", but looks like I overshot it.

richvdh reviewed (commented):
looks generally good, but a few stylistic suggestions

anoadragon453 committed:
Do extra validation on the power level event content before processing

anoadragon453 committed:
Use sets instead of lists

anoadragon453 committed:
Move logging up a touch

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -561,8 +569,6 @@ async def _freeze_room_if_last_admin_is_leaving(
         power_level_content["events"] = {}
 
         # Freeze the room by raising the required power level to send events to 100
```
I think this line needs moving too.

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -446,8 +449,136 @@ def _on_membership_or_invite(
+            # This user is not an admin, ignore them
+            return
+
+        # Filter these users to only those who are actually joined or invited to the room
+        joined_members = [
+            user_id
+            for (event_type, user_id), event in state_events.items()
+            if event_type == EventTypes.Member
+            and event.membership in [Membership.JOIN, Membership.INVITE]
+        ]
+        admin_users = {user for user in admin_users if user in joined_members}
+
+        if len(admin_users) > 1:
+            # There's another admin user in, or invited to, the room
+            return
```
This feels like it's doing a bunch of redundant work. We can bail out as soon as we find an admin member, and we don't need to maintain big lists/sets.

```suggestion
        if any(
            event_type == EventTypes.Member
            and event.membership in [Membership.JOIN, Membership.INVITE]
            and state_key in admin_users
            and state_key != user_id
        ):
            # There's another admin user in, or invited to, the room
            return
```

anoadragon453 replied:
It's actually very rare that I use `any` in my code, but I see the usefulness of it now, thanks!

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -446,8 +449,136 @@ def _on_membership_or_invite(
+        #     "kick": 100,
+        #     "invite": 100
+        # }
+        for key, value in power_level_content.items():
+            # Do not change "users_default", as that key specifies the default power
+            # level of new users
+            if isinstance(value, int) and key != "users_default":
+                power_level_content[key] = 100
+        power_level_content["events"] = {}
```
rather than copying at line 474, why not build a new content object?

```suggestion
        new_content = {}
        for key, value in power_level_content.items():
            # Do not change "users_default", as that key specifies the default power
            # level of new users
            if isinstance(value, int) and key != "users_default":
                value = 100
            new_content[key] = value
        new_content["events"] = {}
```

richvdh replied:
Also, how about PL events were the relevant keys are missing?

anoadragon453 replied:
True, if an event came in that had `user_default` set to `100` then the room wouldn't be very frozen.

The keys we really care about here are `state_default`, `events_default`, `users_default`. Perhaps we should in addition set the first two to `100` explicitly in the new dict?

anoadragon453 replied:
Also sounds like a good idea for another test case!

anoadragon453 replied:
I refactored the existing test in https://github.com/matrix-org/synapse-dinsic/pull/59/commits/a0daf1da647527555d8b18fb358e3bfb6d1d9b35 to create a function that takes a room ID and a dict representing the content of a power levels event, and tries to freeze a room that has that power levels contents.

I then test both the existing case (where the room has the default power levels), and and the case where a power levels event lacking `state_default` and `events_default` keys works (and asserting that those keys now exist and are set to 100).

anoadragon453 committed:
Use any and simplify last admin calculation

anoadragon453 committed:
Move freeze room comment

anoadragon453 committed:
Create a new content dict instead of copying the old one; explicitly set important keys

anoadragon453 committed:
Tests for malformed power level dicts

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -561,20 +559,29 @@ async def _freeze_room_if_last_admin_is_leaving(
+        # Clear out any special-case event power levels
+        new_content["events"] = {}
+
+        # Ensure state_default and events_default keys exist and are 100
+        # Else a lower PL user could potentially send state events that
+        # aren't explicitly mentioned elsewhere in the power level dict
+        new_content["state_default"] = 100
+        new_content["events_default"] = 100
```
we might still have issues with missing `ban`, `kick`, etc keys, each of which default to 50?

giomfo replied:
where is defined this default to 50 for missing `ban`, `kick` and others?

anoadragon453 replied:
It's written in the spec: https://matrix.org/docs/spec/client_server/r0.6.1#m-room-power-levels

![image](https://user-images.githubusercontent.com/1342360/95851851-c17f0480-0d4a-11eb-831c-a0628b7cc6b4.png)

anoadragon453 replied:
We're going to go with setting these to 100 if they're missing for now as we set them to 100 above if they're present anyways.

anoadragon453 replied:
Done in https://github.com/matrix-org/synapse-dinsic/pull/59/commits/6331d52b980fb2ce6aca1d2bf6e36e83a70bea99.

richvdh reviewed (approved):
lgtm otherwise. dunno if the problem is an actual problem here.

anoadragon453 committed:
Set membership event PLs to 100 if they are not specified in the PL event

anoadragon453 committed:
typo
